[LV][VPlan] Implement VPlan-based cost for exit condition.

ElvisWang123 opened the pull request:
This patch tried to model the cost of exit conditions through vplan-based cost model.

* `BranchOnCount` will generate icmp + br.

The branch instruction is already implemented by the VPRegionBlock so we only need to calculate the cost of icmp.

If the VF is same as the trip count of the loop, the cost of the BranchOnCount is free.

There are some tests changes for following reasons.

* Some of the BranchOnCount could be optimized to BranchOnCond, which will generate extract-element + br. And the extract-element is free under some targets.
* Some of the instructions calculated in the exit condition in legacy cost model will be optimized out.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -823,6 +823,29 @@ bool VPInstruction::onlyFirstPartUsed(const VPValue *Op) const {
+  Type *ValTy = Ctx.Types.inferScalarType(getOperand(0));
+
+  switch (getOpcode()) {
+  case VPInstruction::BranchOnCount: {
```
Agree that `BranchOnCond` will generate `extractelement(vec, 0)` and  isn't free for all target. But after this change, the test still change because the extractelement is free under AArch64.

In `clamped_tc_8` the `BranchOnCount` is replaced by `get.active.lane.mask` + `not` + `BranchOnCond`. 
I've tried to implement  the  cost of `get.active.lane.mask` locally but it will impact lots of tests because the cost of this intrinsics is very expensive in AArch64 (2 * element count). And the `clapmed_tc_8` will not even vectorized.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -823,6 +823,29 @@ bool VPInstruction::onlyFirstPartUsed(const VPValue *Op) const {
+    // BranchOnCount will genearte icmp_eq + br instructions and the
+    // cost of branch will be calculated in VPRegionBlock.
+    // If the vector loop only executed once, ignore the cost of the cmp.
+    auto TC = dyn_cast_if_present<ConstantInt>(
+        getParent()->getPlan()->getTripCount()->getUnderlyingValue());
+    if (TC && VF.isFixed() && TC->getSExtValue() == VF.getFixedValue())
+      return 0;
```
The legacy cost model check if the  vectorizer can remove the condition and it will not calculate the cost of   the condition. You can find this in test `fully-unrolled-costs.ll`.

We can remove this after we hoist `unrollByUF` and `optimizeForVFandUF` before estimating instruction cost.

@fhahn Do you have a better method to get the original trip count here?

ElvisWang123 commented:
Ping!
@david-arm @fhahn  I was wondering if you could test this patch and check if it will cause regression under AArch64?
If so, I plan to change the implementation to always adding the cost of `ICmp` at the end of the vector loop to match the behavior of legacy cost model. Thanks!

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7238,45 +7238,6 @@ LoopVectorizationPlanner::precomputeCosts(VPlan &Plan, ElementCount VF,
-                   !ExitInstrs.contains(cast<Instruction>(U));
-          }))
-        continue;
-      ExitInstrs.insert(OpI);
```
It is calculated on each iteration, it won't be hoisted

ElvisWang123 commented:
Gentle ping :)

The potential regression of `clamped_tc_8` is gone.  
If there are still some issues that this patch may cause regression, I will try to always add the cost of icmp to the VPBB (similar to cost of br) to match the legacy cost model.

fhahn reviewed (commented):
Thanks for the update, will try to give this a try this week

ElvisWang123 committed:
[LV][VPlan] Implement VPlan-based cost for exit condition.

This patch tried to model the cost of exit conditions through
vplan-based cost model.

* `BranchOnCount` will generate icmp + br.

The branch instruction is already implemented by the VPRegionBlock so
we only need to calculate the cost of icmp.

If the VF is same as the trip count of the loop, the cost of the
BranchOnCount is free.

This patch is not quite NFC for following reasons.

* Some of the BranchOnCount could be optimized to BranchOnCond, which is
free.
* Some of the instructions calculated in the exit condition in legacy
cost model will be optimized out.

ElvisWang123 commented:
Rebase and gentle ping :)

ElvisWang123 reviewed (commented):
Address comments and based on #152628 that fix some test cases changes.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/reduction-recurrence-costs-sve.ll:
```diff
@@ -453,38 +453,38 @@ define i16 @reduce_udiv(ptr %src, i16 %x, i64 %N) #0 {
-; PRED-NEXT:    [[TMP20:%.*]] = udiv <vscale x 8 x i16> [[WIDE_MASKED_LOAD]], [[BROADCAST_SPLAT]]
-; PRED-NEXT:    [[TMP21:%.*]] = or <vscale x 8 x i16> [[TMP20]], [[VEC_PHI]]
-; PRED-NEXT:    [[TMP16]] = select <vscale x 8 x i1> [[ACTIVE_LANE_MASK]], <vscale x 8 x i16> [[TMP21]], <vscale x 8 x i16> [[VEC_PHI]]
+; PRED-NEXT:    [[WIDE_MASKED_LOAD:%.*]] = call <vscale x 4 x i16> @llvm.masked.load.nxv4i16.p0(ptr [[TMP14]], i32 2, <vscale x 4 x i1> [[ACTIVE_LANE_MASK]], <vscale x 4 x i16> poison)
```
Hmm doesn't https://github.com/llvm/llvm-project/pull/152628 just hide the issue, as it pushes the extract before the `Not`? The extract is still there and only because of the BranchOnCond.  If this really gets code-gen'd, it should probably be accounted for as extracts can be quite expensive on AArch64 and X86 at least.

david-arm commented:
> Yes, the cost of VPInstruction ICmp is not modeled in vplan-based cost model yet.
> 
> But model the cost of ICmp will cause more tests changes and more loops not vectorized.

However, if possible I would still prefer to do this in a different order, i.e.

1. Do preparation work to ensure this PR doesn't regress, then
2. Land this PR.

otherwise if you do this in the opposite order we're knowingly breaking the loop vectoriser for a period of time, which isn't ideal for other users. If the only way to avoid regressions is to add the cost of the icmp to this PR as well, then perhaps that's a necessary evil?

Also, the test `@vectorization_not_profitable_due_to_trunc` is still broken, which needs fixing. The test is very confusing because with this PR we do vectorise.

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
If the cmp is only used by the latch exiting branch I presume it won't generate a vector cmp? In that case do we need to only return the scalar cost
